internal/kubernetes/operator/features/patcher.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2024 MongoDB Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package features
+
+import (
+	"fmt"
+	"strings"
+
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/kube-openapi/pkg/util/sets"
+)
+
+// Patcher is the type that is able to patch Kubernetes objects using a CRD specification.
+type Patcher interface {
+	Patch(crdSpec *v1.JSONSchemaProps, obj runtime.Object) error
+}
+
+// PatcherFunc is a convenience function wrapper around Patcher.
+type PatcherFunc func(crdSpec *v1.JSONSchemaProps, obj runtime.Object) error
+
+func (pf PatcherFunc) Patch(crdSpec *v1.JSONSchemaProps, obj runtime.Object) error {
+	return pf(crdSpec, obj)
+}
+
+// NopPatcher does not patch anything.
+func NopPatcher() Patcher {
+	return PatcherFunc(func(*v1.JSONSchemaProps, runtime.Object) error {
+		return nil
+	})
+}
+
+// UnknownBackupPolicyFrequencyTypesPruner removes backup policy items from a backup policy
+// with unknown frequency types.
+// It inspects the CRD definition to determine supported frequency types.
+func UnknownBackupPolicyFrequencyTypesPruner() Patcher {
+	return PatcherFunc(func(crdSpec *v1.JSONSchemaProps, obj runtime.Object) error {
+		// we are not defensive here as this function assumes the invariant
+		// of a stable CRD definition for a given version of Kubernetes Atlas Operator.
+		frequencyTypePropsEnum := crdSpec.Properties["items"].Items.Schema.Properties["frequencyType"].Enum
+
+		knownFrequencyTypes := sets.NewString()
```

Comment:
do we need to introduce a new lib just for the sets?
sets are just `map[string]struct{}` and seems like you only insert and validate it exists and not of the other more advance operations of a set so I wonder the benefit of another dependency on k8s for so little benefit

Reply:
we have this library from kube already but I am fine with making it explicit.

funny side note: in an upstream or operator PR the review comment would be exactly the reverse :-)

Reply:
fixed

Reply:
>in an upstream or operator PR the review comment would be exactly the reverse :-)

I live by
[A little copying is better than a little dependency.](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=9m28s)

Reply:
ahhh, the good old golang pronouns :-)

Reply:
*proverbs :P